### PR TITLE
HTTP response header regex matching

### PR DESCRIPTION
**Pull request by mitchellrj**
Also enables validating if response headers exist or not.

Defaulting to require all headers to match/not match is something I'm not sure about, but seems like the least worst default. Having multiple headers with the same name is a fairly uncommon scenario.

### CODE REVIEW

**Comment by mitchellrj** [comment]
Any thoughts on this, @SuperQ / @brian-brazil? :)

File: config/config.go [hunk 1]
```diff
@@ -50,20 +50,28 @@ type Module struct {
 	DNS     DNSProbe      `yaml:"dns,omitempty"`
 }
 
+type HTTPHeaderMatch struct {
+	Header   string `yaml:"header,omitempty"`
+	Pattern  string `yaml:"pattern,omitempty"`
+	Required bool   `yaml:"pattern,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: pattern?

File: CONFIGURATION.md [hunk 2]
```diff
@@ -66,6 +66,22 @@ The other placeholders are specified separately.
...
+        pattern: <regex>
+        # Also fail if the header is missing from the response.
+        required: <bool>
+      ... ]
+
+  # Probe fails if response does not header matches regex.
+  fail_if_header_not_matches_regexp:
+    [ - header: <header-name>
+        pattern: <regex>
+        # Also fail if the header is missing from the response.
```
Comment (brian-brazil) [inline CONFIGURATION.md]: How would you assert that a header must be missing?

File: CONFIGURATION.md [hunk 3]
```diff
@@ -66,6 +66,22 @@ The other placeholders are specified separately.
   fail_if_not_matches_regexp:
```
Comment (brian-brazil) [inline CONFIGURATION.md]: This needs clarification that it's for the body

File: prober/http.go [hunk 4]
```diff
@@ -169,9 +209,14 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 			Help: "Returns the version of HTTP of the probe response",
 		})
 
-		probeFailedDueToRegex = prometheus.NewGauge(prometheus.GaugeOpts{
+		probeFailedDueToHeaderRegex = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_failed_due_to_header_regex",
```
Comment (brian-brazil) [inline prober/http.go]: I think one metric for this is fine, the debug logs will always have the full details

File: prober/http.go [hunk 5]
```diff
@@ -38,13 +38,53 @@ import (
...
+				return false
+			}
+			for _, value := range headerValues {
+				if re.MatchString(value) {
+					level.Error(logger).Log("msg", "Header matched regular expression", "regexp", key, value, matcher.Pattern)
+					return false
+				}
+			}
+		} else if matcher.Required {
+			return false
```
Comment (brian-brazil) [inline prober/http.go]: Log when you return false so the user can know why it failed

**Comment by mitchellrj** [comment]
I've been away on holiday, I shall pick it up again this week, thanks.

**Comment by brian-brazil** [comment]
Did you make any progress since your holiday?

**Comment by brian-brazil** [comment]
Obsoleted by #419

**Pull request closed without merging**